src/main/java/ecdar/backend/BackendDriver.java

```diff
@@ -166,18 +138,101 @@ public void onNext(Empty value) {
 
             @Override
             public void onError(Throwable t) {
+                newConnection.close();
+                startedEngineConnections.remove(newConnection);
             }
 
             @Override
             public void onCompleted() {
-                setConnectionAsAvailable(newConnection);
+                if (startedEngineConnections.contains(newConnection)) setConnectionAsAvailable(newConnection);
             }
         };
 
         newConnection.getStub().withDeadlineAfter(responseDeadline, TimeUnit.MILLISECONDS)
                 .updateComponents(componentsBuilder.build(), observer);
     }
 
+    /**
+     * Starts a process, creates an EngineConnection to it, and returns that connection
+     *
+     * @param engine to run in the process
+     * @return an EngineConnection to a local engine running in a Process or null if all ports are already in use
+     */
+    private EngineConnection startAndGetConnectionToEngineProcess(final Engine engine) {
+        long port;
+        try {
+            port = SocketUtils.findAvailableTcpPort(engine.getPortStart(), engine.getPortEnd());
+        } catch (IllegalStateException e) {
+            // All ports specified for engine are already used for running engines
+            return null;
+        }
+
+        // Start local process of engine
+        Process p;
+        int attempts = 0;
+
+        do {
+            attempts++;
+            ProcessBuilder pb = new ProcessBuilder(engine.getEngineLocation(), "-p", "127.0.0.1:" + port);
+
+            try {
+                p = pb.start();
+            } catch (IOException ioException) {
+                Ecdar.showToast("Unable to start local engine instance");
+                ioException.printStackTrace();
+                return null;
+            }
+        } while (!p.isAlive() && attempts < maxRetriesForStartingEngineProcess);
+
+        ManagedChannel channel = startGrpcChannel("127.0.0.1", port);
+        EcdarBackendGrpc.EcdarBackendStub stub = EcdarBackendGrpc.newStub(channel);
+        return new EngineConnection(engine, channel, stub, p);
+    }
+
+    /**
+     * Creates and returns an EngineConnection to the remote engine
+     *
+     * @param engine to connect to
+     * @return an EngineConnection to a remote engine or null if all ports are already connected to
+     */
+    private EngineConnection startAndGetConnectionToRemoteEngine(Engine engine) {
```

Comment:
I kind of have the same comment for this function as with ``startAndGetConnectionToEngineProcess``.